[refactor]: enhance compiler (#70)
soramitsu/scale-codec-js-library: .changeset/two-pumpkins-wait.md

---
'@scale-codec/definition-compiler': major
'@scale-codec/definition-runtime': major
---

> **refactor**: use local opaque type pattern, remove `type-fest` dep (and `Opaque` re-export); workaround enum issues with `EnumBox` type; update generation, generate less code

Reviewer: ```suggestion
**refactor**: use local opaque type pattern, remove `type-fest` dep (and `Opaque` re-export); implement workaround enum issues with `EnumBox` type; update generation to produce less code
```

> pattern

is it just a single pattern?

> dep

dependency?

Author: > is it just a single pattern?

Yes

> dep

Yes

Author: Changed to

> use local opaque type pattern, remove `type-fest` dependency (and `Opaque` re-export accordingly); work around enum issues with `EnumBox` type; update generation to produce less code

> **refactor**: use local opaque type pattern, remove `type-fest` dependency (and `Opaque` re-export accordingly); work around enum issues with `EnumBox` type; update generation to produce less code

Reviewer: ```suggestion
**refactor**: use the local opaque type pattern, remove `type-fest` dependency (and `Opaque` re-export accordingly); work around enum issues with `EnumBox` type; update generation to produce less code
```

#### Opaque types

> Each generated codec type is **opaque** - it can be instantiated only through the codec factory or by explicit `as` conversion. It **was** implemented with `Opaque` type from `type-fest` library. Now it is implemented with the following technique:

Reviewer: ```suggestion
Each generated codec type is **opaque**, which means it can be instantiated only through the codec factory or by an explicit `as` conversion. It used to be implemented via `Opaque` type from `type-fest` library. Now it is implemented in the following way:
```

```ts
declare const __opaqueTag: unique symbol

type LocalOpaque<Tag, T> = { [__opaqueTag]: Tag } & T

// ...some type
type VecU32 = LocalOpaque<'VecU32', number[]>
```

> This approach allows to define types with same name in different generated modules and not to worry about their intersections.

Reviewer: ```suggestion
With this approach it is possible to define types with the same name in different generated modules without worrying about naming conflicts.
```

> **A consumer shouldn't update their code**, unless they used `Opaque` type on their own.

Reviewer: ```suggestion
**No code updates are required unless `Opaque` type was used.**
```

#### Enum issues and `EnumBox`

> Previous enum generation approach had a set of drawbacks when it came to large namespaces and circular references. TypeScript computed wrong types for enum factories and frequently failed to resolve types in general because of circuits.

Reviewer: ```suggestion
The previous enum generation approach had a set of drawbacks when it came to large namespaces and circular references. TypeScript computed wrong types for enum factories and frequently failed to resolve types in general because of circuits.
```

Now all enum codecs are based on `EnumBox<V>` type:

```ts
type EnumBox<V extends VariantAny> = {
  enum: V
}
```

> With this approach, TypeScript is able to handle abovementioned problems, but it has a runtime drawback - **each enum is now wrapped** into `EnumBox`.

Reviewer: ```suggestion
With this approach, TypeScript is able to handle abovementioned problems. However, it has a runtime drawback: each enum is now wrapped into `EnumBox`.
```

> A consumer should add `.enum` property accessor in any place they work with enums:

Reviewer: ```suggestion
You should add `.enum` property accessor in any place you work with enums:
```
